### DISCUSSION TITLE
Implement state-feedback control on nonlinear aeroelastic model

### DISCUSSION THREAD

**Discussion by Alanpwu** [ImperialCollegeLondon/sharpy#149] [Q&A] [answered]
Hi, 

I had modelled an aeroelastic wing and obtained the state-space(A,B,C,D) of the linearised system. Is it possible in SHARPy where i can obtain the states value (\eta \zeta)  can calculate the control surface deflection all within the DynamicCoupled Solvers? I am doing this for gust load alleviation purposes. 

Thank you!

**Comment by ngoiz** [comment, marked as answer]
Hi,

I'm not sure if I exactly follow what you are trying to achieve. I assume that by value of the states you refer to a time domain simulation, if so, I would recommend you simulate the linear time domain system externally to SHARPy (i.e. using `scipy.signal.dlsim()` or Matlab). The typical output of a linear time domain simulation from these functions is the output `y` and also the states `x` of the system. In SHARPy, the `DynamicCoupled` solver is intended for nonlinear time-domain simulations.

Hope this helps

**Reply by Alanpwu** [reply]
Hi, 

I'm sorry the description for my problem is not clear.  My goal is to perform gust load alleviation of an aircraft wing and wish to perform state-feedback control within SHARPy's DynamicCoupled solver, as I have set the velocity_field_generator to 'GustVelocityField'. 

So I have the linear state-space(A,B,C,D) around a fixed geometry now for the purpose of calculating a control gain K = lqr(A,B,Q,R);

Is it achievable in DynamicCoupled that at each time step (outputing the time-domain solutions) --> (calculate the control surface deflection u = -Kx(for my case)) -->  (feed the control surface deflection to next time step)? 

Thank you!

**Reply by ngoiz** [reply]
Hmm, the short answer is no. The controller currently implemented for the `DynamicCoupled` solver is intended for single-input single-output systems. I understand your controller feeds back the state and at this moment this is not implemented. 

It wouldn't be too hard to implement though. Most likely, the easiest would be to use the SHARPy UDP Hardware-in-the-loop capabilities [https://ic-sharpy.readthedocs.io/en/master/includes/io/network_interface/NetworkLoader.html](https://ic-sharpy.readthedocs.io/en/master/includes/io/network_interface/NetworkLoader.html). In your case, you would need to specify all your state variables for SHARPy to send over UDP. In your client, you can then process that with your gain and return the control surface deflection to SHARPy. You can see some example of a server-client set-up in the SHARPy tests (`/tests/io/sample_udp_inout/`)

This is the least code-intensive way of doing so, the alternative being to do this within the SHARPy controllers structure which may be more involved if you are not experienced with it.

**Reply by Alanpwu** [reply]
Thank you! 
What if the the state-space is in reduced order? Since the DynamicCoupled only provide time-domain solution in nodal form.

**Reply by ngoiz** [reply]
I believe that is more for you to think about. Think of SHARPy as your real system where you are sensing the outputs/state via sensors. You may need to manipulate that information to feed it to your controller

**Reply by ngoiz** [reply]
Hi, giving it another thought it would be very cumbersome to send the whole state via UDP. A colleague of mine was attempting this and it proved to be problematic. You may want to explore Kalman filters or some other estimation technique to recreate the state based on a few measurements that you can extract from the nonlinear time domain simulation in SHARPy. This may be of interest for that:  https://doi.org/10.2514/6.2021-1392

**Reply by Alanpwu** [reply]
Thank you so much! 

There is another issure. When i tried to output the tip node psi_dot(4,2,0) (4th element, 2nd node, 1st index), as i wrote position [4,2,0] or position[4,2] & index: 1.    in the .yaml. 

The result is always a [3x3x3] matrix. How can i configure the position or index so that SHARPy only output one value? 

Thank you!

**Reply by ngoiz** [reply]
You spotted a limitation of the system right there. Not sure how but we never used UDP to export the CRV. So, (without setting precedent, I had a spare minute :) ) I have quickly given it the functionality to export CRVs. Hopefully it will work. Please checkout the branch `fix_udp_crv` to use this.

So what you need to do is edit your `yaml` file for the CRV variables as follows:
```
- name: 'psi_dot'
  inout: 'out'
  position: 4
  index: [0, 2]
  var_type: 'elem'
```

where `position` refers to the element, and `index` is now a 2-list where the first index is the node index (0, 1 or 2) and then the second element is the vector component (again, 0, 1, 2).

Hopefully this works. Please let me know so we can roll it out to the release version.

**Reply by Alanpwu** [reply]
where i find the branch fix_udp_crv?

**Comment by rafapalacios** [comment]
https://github.com/ImperialCollegeLondon/sharpy/tree/fix_udp_crv
